src/Features/Core/Portable/Completion/CompletionHelper.cs

```diff
@@ -0,0 +1,601 @@
+                        diff = PatternMatch.CompareCase(match1, match2);
+                        if (diff != 0)
+                        {
+                            return diff;
+                        }
+                        break;
+
+                    case PatternMatchRuleKind.Punctuation:
+                        diff = PatternMatch.ComparePunctuation(match1, match2);
+                        if (diff != 0)
+                        {
+                            return diff;
+                        }
+                        break;
+
+                    case PatternMatchRuleKind.ArgumentName:
+                        // argument names are ordered last
+                        if (item1.IsArgumentName && !item2.IsArgumentName)
+                        {
+                            return 1;
+                        }
+                        else if (item2.IsArgumentName && !item1.IsArgumentName)
+                        {
+                            return -1;
+                        }
+                        break;
+
+                    case PatternMatchRuleKind.Preselect:
+                        // preselected items are ordered first
+                        if (item1.Preselect && !item2.Preselect)
+                        {
+                            return -1;
+                        }
+                        else if (item2.Preselect && !item1.Preselect)
+                        {
+                            return 1;
+                        }
+                        break;
+                }
+            }
+
+            return 0;
+        }
+
+        /// <summary>
+        /// Returns true if item1 and item2 are similar enough that only one should be shown in the completion list; otherwise, false.
+        /// </summary>
+        public bool ItemsMatch(CompletionItem item1, CompletionItem item2)
+        {
+            foreach (var rule in _rules.ItemMatchRules)
+            {
+                switch (rule.Kind)
+                {
+                    case ItemMatchRuleKind.FilterSpan:
+                        if (item1.FilterSpan != item2.FilterSpan)
+                            return false;
+                        break;
+
+                    case ItemMatchRuleKind.SortText:
+                        if (item1.SortText != item2.SortText)
+                            return false;
+                        break;
+
+                    case ItemMatchRuleKind.Glyph:
+                        var se = SpecialException(item1, item2);
+                        if (se != null)
+                        {
+                            return se.Value;
+                        }
+
+                        if (!object.Equals(item1.Glyph, item2.Glyph))
+                            return false;
+                        break;
+                }
+            }
+
+            return true;
+        }
+
+        private bool? SpecialException(CompletionItem item1, CompletionItem item2)
+        {
```

Comment:
This seems awful.  It shouldn't be the case that this layer is even aware of language names.  All language specific behavior should always be in a library provided by that language. This really violates that pattern we've been following everywhere else.